### PR TITLE
Immediate Execution of `Commands` on a `&mut World`

### LINKED ISSUE #6184: Immediate Execution of Commands on `&mut World`
## What problem does this solve or what need does it fill?

Currently, to write a test for a command, we have 2 options:
1. Either create a minimal app with a system that uses the command, or
2. Use `CommandQueue` with a `World` to execute commands "in place"

I've noticed we already use option 2 in Bevy code:

```rust
fn remove_resources() {
    let mut world = World::default();
    let mut queue = CommandQueue::default();
    {
        let mut commands = Commands::new(&mut queue, &world);
        commands.insert_resource(W(123i32));
        commands.insert_resource(W(456.0f64));
    }

    queue.apply(&mut world);
    assert!(world.contains_resource::<W<i32>>());
    assert!(world.contains_resource::<W<f64>>());

    {
        let mut commands = Commands::new(&mut queue, &world);
        // test resource removal
        commands.remove_resource::<W<i32>>();
    }
    queue.apply(&mut world);
    assert!(!world.contains_resource::<W<i32>>());
    assert!(world.contains_resource::<W<f64>>());
}
```

My proposal is that we simplify this by adding a function to execute commands directly on a world.

## What solution would you like?

I have a minimal working implementation of this in my own project:

```rust
use bevy::prelude::*;

use bevy::ecs::system::CommandQueue;

///
/// Trait to execute [`Commands`] on a [`World`].
///
pub trait Execute {
    fn execute<F: FnOnce(&World, Commands) -> R, R>(self, f: F) -> R;
}

///
/// Any mutable reference to a [`World`] may be used to execute [`Commands`] on it.
///
impl Execute for &mut World {
    fn execute<F: FnOnce(&World, Commands) -> R, R>(self, f: F) -> R {
        let mut queue = CommandQueue::default();
        let commands = Commands::new(&mut queue, self);
        let result = f(self, commands);
        queue.apply(self);
        result
    }
}

///
/// Any mutable reference to an [`App`] may be used to execute [`Commands`] on its [`World`].
///
impl Execute for &mut App {
    fn execute<F: FnOnce(&World, Commands) -> R, R>(self, f: F) -> R {
        self.world.execute(f)
    }
}
```

This allows you to just do this anywhere you have `&mut World` access:
```rust
world.execute(|_world, mut commands| {
    /* ... */
});
```

## What alternative(s) have you considered?

See above for current options.

## Additional context

I tried making this a direct pull request. But as I was implementing it, I wasn't sure where this should go. I was thinking of putting `Execute` in `bevy_ecs/src/system/commands/mod.rs`, but I wasn't sure if it's smart to import `Commands` in `bevy_app/src/lib.rs`

If we're ok with the implementation I've proposed above, and importing `Commands` in Bevy app, then I can just turn this into a PR.
Also I was internally debating guarding this behind a feature flag, so that it's only available for tests/examples. Still not sure if that's wise.

### PR DESCRIPTION
**Pull request by Zeenobit**
# Objective

Fixed #6184 

## Solution

This PR adds an `Execute` trait which is implemented for `&mut World` and `&mut App`. It allows the user to do this:

```rust
let mut world = World::default();
let _ = world.execute(|_world, mut commands| {
    /* ... */
});
```

### CODE REVIEW

**Comment by mockersf** [comment]
As you mention in your doc comment, this is a very inefficient way to execute a command when you have a world available.

If this is only to be used for test, I would rather it being not available outside of tests

**Comment by Zeenobit** [comment]
> As you mention in your doc comment, this is a very inefficient way to execute a command when you have a world available.
> 
> If this is only to be used for test, I would rather it being not available outside of tests

I considered doing this initially. There is a couple of problems with this:

1. We can't just gate it behind `#[cfg(test)]`, as it would also be useful for examples. So ideally it should be locked a feature of its own, but that's just extra build complexity for questionable benefit.
2. I've found this to be useful for debug UI code, such as with egui, where I want to have a button which executes a bunch of commands when I have `&mut World` access. So it's also useful for general diagnostics at runtime.

For these reasons, I decided keeping it available but with a warning in the comments is probably the best option.

Also, `Execute` is not imported with preludes. So that's an extra layer of "protection".

File: crates/bevy_ecs/src/system/commands/mod.rs [hunk 1]
```diff
@@ -907,13 +907,50 @@ impl Command for LogComponents {
...
+pub trait Execute {
+    fn execute<F: FnOnce(&World, Commands) -> R, R>(self, f: F) -> R;
+}
+
+impl Execute for &mut World {
+    /// Creates a new [`Commands`] instances and passes it to the given function, then executes it on this [`World`].
+    ///
+    /// # Usage
+    ///
+    /// This is not an efficient method of executing commands. However, it can be a convenient tool for
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/commands/mod.rs:920]: This should explain why it's not efficient, and the more efficient alternatives.

**Comment by alice-i-cecile** [comment]
> If this is only to be used for test, I would rather it being not available outside of tests

This came up before in my other PRs for testing utilities. I think that the trait (that's not in the prelude) is a reasonable approach. I would also support the creation of a `bevy_testing` crate to stick this sort of thing in, which would improve discoverability.

**Comment by Zeenobit** [comment]
> This came up before in my other PRs for testing utilities. I think that the trait (that's not in the prelude) is a reasonable approach. I would also support the creation of a `bevy_testing` crate to stick this sort of thing in, which would improve discoverability.

I like this idea, but I would suggest naming it some more generic like `bevy_diagnostics`. My rationale is that a lot of stuff like this that's useful for testing would also be useful for stuff like cheats, debug tools, or automation. If we name it `bevy_testing`, it'd imply it's *only* useful for testing.

**Comment by alice-i-cecile** [comment]
Well, we *have* `bevy_diagnostic`. @mockersf, I'd be okay with testing utilities going in there!

**Comment by Zeenobit** [comment]
Let me know if we want to move the entire implementation into a separate file `execute.rs` under `bevy_diagnostic`.

**Comment by alice-i-cecile** [comment]
Yes please :)

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Awesome. Can you add an example for this? Maybe in the root tests directory? I would suggest you link it from the Commands docs but we can't, because circular dependencies.

**Comment by Zeenobit** [comment]
> Awesome. Can you add an example for this? Maybe in the root tests directory? I would suggest you link it from the Commands docs but we can't, because circular dependencies.

I added an example under `how_to_test_commands.rs`. It's hard to come up with a good use case for using a custom command without really explaining a system which might need it. I figured navigation might the best example, but let me know if you have better ideas. And for all regular commands, most of them are accessible through `&mut World` anyways.

Also, I'm not sure if it was just me, but an issue I discovered while testing:

`world.spawn(())` seems to panic with "index out of bounds: the len is 0 but the index is 0". But `commands.spawn(())` seems fine.

File: crates/bevy_diagnostic/src/execute.rs [hunk 2]
```diff
@@ -24,7 +24,7 @@ impl Execute for &mut World {
     ///
     /// let mut world = World::default();
     /// let entity = world.execute(|_world, mut commands| {
-    ///     commands.spawn().id()
+    ///     commands.spawn(()).id()
```
Comment (alice-i-cecile) [inline crates/bevy_diagnostic/src/execute.rs:27]: This should use `spawn_empty` instead.

File: crates/bevy_diagnostic/src/execute.rs [hunk 3]
```diff
@@ -0,0 +1,46 @@
+use bevy_app::App;
+use bevy_ecs::prelude::{Commands, World};
+use bevy_ecs::system::CommandQueue;
+
+/// Extension trait used to execute some [`Commands`] immediately on a [`World`].
```
Comment (alice-i-cecile) [inline crates/bevy_diagnostic/src/execute.rs:5]:
```suggestion
/// Extension trait used to execute [`Commands`] immediately on a [`World`].
```

File: crates/bevy_diagnostic/src/execute.rs (hunk 2 above)
Comment (Zeenobit) [reply crates/bevy_diagnostic/src/execute.rs:27]: Ah that explains a lot! Thanks!

**Review by alice-i-cecile — APPROVED** [review]

**Comment by Zeenobit** [comment]
Today I realized I accidentally deleted the repository for this. 😢 
I'll open a new PR in future.

**Pull request closed without merging**